### PR TITLE
Updated classmap generator

### PR DESCRIPTION
**Pull request by janvernieuwe**
TODO:
* (DONE) Update tests (currently failing due to the changes)
* (DONE) Update documentation
* (DONE) Handle existing files
* (Not in scope) Add functionality to make a type hint overridable (ie SomeISOStringDefinitionThatIsNotATypeClass to string)

### CODE REVIEW

File: src/Phpro/SoapClient/CodeGenerator/Config/Config.php [hunk 1]
```diff
@@ -45,7 +45,7 @@ class Config implements ConfigInterface
      * @var array
      */
     protected $soapOptions = [
-        'trace' => false,
+        'trace'      => false,
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Config/Config.php]: dude ... ;)

File: src/Phpro/SoapClient/CodeGenerator/Config/Config.php [hunk 2]
```diff
@@ -80,12 +95,14 @@ class Config implements ConfigInterface
     public function __construct(string $wsdl = '', string $destination = '')
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Config/Config.php]: The destination parameter doesn't make a lot of sense anymore, lets drop it?

File: src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php [hunk 3]
```diff
@@ -63,11 +74,15 @@ protected function configure()
...
     {
+        $this->input = $input;
+        $this->output = $output;
+
         $configFile = $input->getOption('config');
         if (!$configFile || !$this->filesystem->fileExists($configFile)) {
             throw InvalidArgumentException::invalidConfigFile();
         }
 
+        /** @noinspection PhpIncludeInspection */
```
Comment (veewee) [inline src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php]: No need for this

File: src/Phpro/SoapClient/CodeGenerator/Config/Config.php (hunk 2 above)
Comment (janvernieuwe) [reply src/Phpro/SoapClient/CodeGenerator/Config/Config.php]: I had it there for BC, I'll remove it

File: src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php (hunk 3 above)
Comment (janvernieuwe) [reply src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php]: Ok, phpstorm whines that it cannot resolve the include then.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php [hunk 4]
```diff
@@ -61,10 +80,10 @@ private function assembleClassMap(TypeMap $typeMap, string $linefeed, string $in
         $classMap = [];
         foreach ($typeMap->getTypes() as $type) {
             $classMap[] = sprintf(
-                '%snew ClassMap(\'%s\', \\%s::class),',
+                '%snew ClassMap(\'%s\', %s::class),',
                 $indentation,
                 $type->getXsdName(),
-                $type->getFullName()
+                'Type\\'.$type->getName()
```
Comment (GaryJones) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php]: Why would this be hard-coded to `Type`?

Here's what my current generated classpmap looks like:

screenshot 2018-01-04 13 31 15

File: src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php [hunk 5]
```diff
@@ -76,8 +90,127 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+    protected function generateClassmap(FileGenerator $file, ClassMapGenerator $generator, TypeMap $typeMap, $path)
+    {
+        $code = $generator->generate($file, $typeMap);
+        $this->filesystem->putFileContents($path, $code);
+    }
+
+    /**
+     * Try to create a class for a type.
+     * When a class exists: try to patch
+     * If patching the old class does not wor: ask for an overwrite
```
Comment (GaryJones) [inline src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php]: Typo: `wor` => `work`.

File: src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php [hunk 6]
```diff
@@ -76,8 +90,127 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+
+            // Ask if a class can be overwritten if it contains errors
+            if (!$this->askForOverwrite()) {
+                $this->output->writeln(sprintf('Skipping %s', $path));
+
+                return false;
+            }
+        }
+
+        // Try to create a blanco class:
```
Comment (GaryJones) [inline src/Phpro/SoapClient/Console/Command/GenerateClassmapCommand.php]: What's a "blanco" class?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php [hunk 7]
```diff
@@ -33,17 +35,34 @@ public function canAssemble(ContextInterface $context): bool
...
+            [
+                'name' => $context->getName(),
+            ]
+        );
         $file = $context->getFile();
+        $file->setClass($class);
+        $file->setNamespace($context->getNamespace());
         $typeMap = $context->getTypeMap();
+        $typeNamespace = $typeMap->getNamespace();
+        $file->setUse($typeNamespace, preg_match('/Type$/', $typeNamespace) ? null : 'Type');
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php]: This also maches stuff like "SoapType" which will cause an error

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php (hunk 4 above)
Comment (janvernieuwe) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/ClassMapAssembler.php]: This is an alias for the use.
Just to make things look a little nicer.

**Pull request merged**